ingestors/npm.go

```diff
@@ -1,154 +1,116 @@
 			break
 		}
-	}
+		results = append(results, lastResults...)
 
-	for {
-		if changes.Next() {
-			var doc NPMChangeDoc
-			if err := changes.ScanDoc(&doc); err != nil {
-				log.WithFields(log.Fields{"seq": changes.Seq(), "id": changes.ID()}).Fatal(err)
-			}
-			var latestVersion string
-			var latestTime time.Time
-			for k, v := range doc.Time {
-				if k != "modified" && k != "created" && len(v) > 0 {
-					if t, err := time.Parse(time.RFC3339, v); err == nil {
-						if t.After(latestTime) {
-							latestTime = t
-							latestVersion = k
-						}
-					}
-				}
-			}
-			if latestVersion != "" {
-				discoveryLag := time.Since(latestTime)
-				results <- data.PackageVersion{
-					Platform:     "npm",
-					Name:         doc.Name,
-					Version:      latestVersion,
-					CreatedAt:    latestTime,
-					DiscoveryLag: discoveryLag,
-				}
-				since = changes.Seq()
-				if _, err := setBookmark(ingestor, since); err != nil {
-					log.WithFields(log.Fields{"ingestor": ingestor.Name()}).Fatal(err)
-				}
-			}
-		} else {
-			log.WithFields(log.Fields{"ingestor": ingestor.Name(), "error": changes.Err()}).Error(fmt.Sprintf("Reconnecting in %d seconds.", ReconnectRetryDelaySeconds))
-			time.Sleep(ReconnectRetryDelaySeconds * time.Second)
-			couchDb = ingestor.couchClient.DB(NPMRegistryDatabase)
-			options := getOptionsForChangesFeed(since)
-			changes = couchDb.Changes(context.Background(), options)
-			if err = changes.Err(); err != nil {
-				log.WithFields(log.Fields{"ingestor": ingestor.Name()}).Fatal(err)
-			}
+		if lastSequence > currentSequence {
+			currentSequence = lastSequence
 		}
 	}
+
+	if _, err := setBookmark(ingestor, strconv.FormatInt(currentSequence, 10)); err != nil {
+		log.WithFields(log.Fields{"ingestor": ingestor.Name()}).Fatal(err)
+	}
+
+	return results
 }
 
-func getCouchClient() *kivik.Client {
-	kivikClient, err := kivik.New("couch", NPMRegistryHostname)
+func (ingestor *NPM) getPage(sequence int64) (int64, []data.PackageVersion) {
+	var results []data.PackageVersion
+
+	// The header enables the new API changes and can be removed May 29th, 2025:
+	// https://github.blog/changelog/2025-02-27-changes-and-deprecation-notice-for-npm-replication-apis/
+	response, err := depperGetUrl(
+		fmt.Sprintf("%s?since=%d&limit=%d", npmIndexUrl, sequence, perPage),
+		map[string]string{"npm-replication-opt-in": "true"},
+	)
+	if err != nil {
+		log.WithFields(log.Fields{"ingestor": ingestor.Name(), "error": err}).Error()
+		return sequence, results
+	}
+	defer response.Body.Close()
+
+	body, _ := io.ReadAll(response.Body)
+	sequenceList, _, _, _ := jsonparser.Get(body, "results")
+	_, _ = jsonparser.ArrayEach(sequenceList, func(value []byte, dataType jsonparser.ValueType, offset int, err error) {
+		name, _ := jsonparser.GetString(value, "id")
+		seq, _ := jsonparser.GetInt(value, "seq")
+
+		// The new NPM feed only provides names and sequences, so we don't get Version, CreatedAt or DiscoveryLag.
```

Comment:
Losing versions is going to have some downstream effects as some of the npm packages that have a gajillion releases (I'm looking at you caniuselite) will now have to pull the full version information every time a new release is published and that has caused database thrashing in the past.

Reply:
I thought we had set a `lock` option on PackageManagerDownloadWorker to delete duplicate jobs, but it seems like not: https://github.com/librariesio/libraries.io/blob/main/app/workers/package_manager_download_worker.rb#L8

An alternate strategy could be:

* for each package event in the response:
  * get the "rev" from the package event
  * check the package on the registry 
  * does the package metadata have the matching "rev" value?
    * if **yes**, then pull the latest version from the metadata and enqueue that version only?
    * if **no**, ... I'm not sure what to do. The mismatch could be bc the metadata is behind **or** ahead.

Reply:
Oh, I''m forgetting that the TTLer feature has a default of "24 hours". So once it publishes a package without version (e.g. "npm/foo") it'll set a `"depper:ingest:npm:foo:"` key with an expiry of 24 hours, blocking any more events for that package.

So it's good that we throttle them, but also probably too restrictive? Maybe "1 hour" for NPM?

Reply:
🤔  Let's talk live about options tomorrow

Reply:
We looked through the code live and we no longer save all versions when doing a version-less sync so I think this is safe